neurogym/wrappers/components/layers/base.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, ClassVar
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Iterable
+
+    import numpy as np
+    import torch
+    from torch import nn
+
+
+def to_numpy(tensor: torch.Tensor) -> np.ndarray:
+    """A convenience method for converting a tensor into a NumPy array.
+
+    Args:
+        tensor: A PyTorch tensor
+
+    Returns:
+        A NumPy array.
+    """
+    return tensor.detach().clone().squeeze().cpu().numpy()
+
+
+class LayerMonitorBase(ABC):
+    layer_type: type[nn.Module]
+    probes: ClassVar
+
+    @staticmethod
+    def get_monitor(
+        layer: nn.Module,
+        hook: Callable,
+        populations: set[str] | None = None,
+    ) -> LayerMonitorBase:
+        """Mapping from Torch layer type to layer monitor.
+
+        Args:
+            layer: The layer being monitored.
+            hook: A function to register as a forward hook.
+            populations: Neuron populations to monitor.
+
+        Returns:
+            A dictionary mapping layer types to layer monitors.
+
+        Raises:
+            NotImplementedError: Raised if a monitor for this layer type has not been implemented.
+        """
+        cls = layer.__class__
+        for layer_monitor in LayerMonitorBase.__subclasses__():
+            layer_type = layer_monitor.layer_type
+            if cls is layer_type:
+                return layer_monitor(layer, hook, populations=populations)  # type: ignore[abstract]
+
+        msg = f"It seems that a monitor for layers of type '{cls}' has not been implemented yet."
+        raise NotImplementedError(msg)
+
+    def __init__(
+        self,
+        layer: nn.Module,
+        hook: Callable,
+        populations: str | Iterable | None,
+    ):
+        self.layer = layer
+
+        cls = self.__class__
+
+        if cls.probes is None:
+            cls.probes = {}
+
+        if populations is None:
+            populations = set(cls.probes.keys())
+
+        elif isinstance(populations, str):
+            populations = {populations}
+
+        else:
+            populations = set(populations)
+
+        self.populations = populations.intersection(set(cls.probes.keys()))
```

Comment:
Can we remove some of the empty spaces here? It looks very spread out to me.

Reply:
Sure, although to me it looks cramped without the spaces...😂

Reply:
I'm surprised that Ruff doesn't have something to say about that, it's usually very opinionated about formatting.